Add support for s390x installer

tmh1999 opened the pull request.

tmh1999 commented:
@jirutka Please review these so I could port changes to https://github.com/alpinelinux/aports/pull/4062.

jirutka commented on initramfs-init.in:
```diff
@@ -381,6 +383,31 @@ fi
 # zpool reports /dev/zfs missing if it can't read /etc/mtab
 ln -s /proc/mounts /etc/mtab
 
+if echo $@ | grep -q "dasd" || echo $@ | grep -q "qeth" ; then
```
This is _very_ bad approach. It may fail when `$1` is happened to be something like `-e`, `-n`, … `$@` should be quoted, otherwise it wouldn’t make much sense to use `$@` instead of `$*`.

What is `dasd` and `qeth`? Do you want to find exact match only or any string that contains one of these as a substring?

jirutka commented on initramfs-init.in:
```diff
@@ -381,6 +383,31 @@ fi
+	done
+	_tmpfile="$(mktemp)"
+	echo $@ | tr A-Z a-z | xargs -n 1 | grep '=' > "$_tmpfile"
+	. "$_tmpfile" > /dev/null 2>&1
```
Is it really needed to pollute global namespace with arbitrary variables…?!

tmh1999 replied:
I updated the code to use current getopts/myopts code instead of this.

tmh1999 replied:
Thanks.

jirutka commented on initramfs-init.in:
```diff
@@ -345,6 +345,26 @@ mount -t devpts -o gid=5,mode=0620,noexec,nosuid devpts /dev/pts
 mount -t tmpfs -o nodev,nosuid,noexec shm /dev/shm
 
+if [ -n "$dasd" ]; then
+	for mod in dasd_mod dasd_eckd_mod dasd_fba_mod; do modprobe $mod; done
```
Please don’t write `for` loop as one-line.

tmh1999 replied:
Updated. Thanks.

ncopa commented on initramfs-init.in:
```diff
@@ -269,7 +269,7 @@ set -- $(cat /proc/cmdline)
 	cryptdiscards debug_init dma init_args keep_apk_new modules ovl_dev pkgs quiet
 	root_size root usbdelay ip alpine_repo apkovl alpine_start splash blacklist
-	overlaytmpfs rootfstype rootflags nbd resume"
+	overlaytmpfs rootfstype rootflags nbd resume nameserver s390x_installer s390x_net dasd"
```
where comes `nameserver` from? Maybe add it in separate commit if it is supposed to be there.

tmh1999 replied:
`nameserver` and `s390x_installer` are meant to be in the second commit. I move them there now.

ncopa commented on Makefile:
```diff
@@ -62,15 +62,17 @@ FULL_VERSION    := $(VERSION)
 
+ifeq ($(ARCH), s390x)
+CONF_FILES	+= features.d/qeth.modules features.d/dasd_mod.modules
+_dummy		:= $(shell ${SED} -i -e 's:features=":features="qeth dasd_mod :g' mkinitfs.conf)
```
This will add `qeth` and `dasd_mod` to mkinitfs.conf every time you run `make`. I think we may need to solve this differently, and without `sed -i ...`

tmh1999 replied:
Yes I was skeptical about this from the begining but couldn't think of other way.
Now I keep features.d/* part here but move mkinitfs.conf part to APKBUILD in aports.

tmh1999 commented on Makefile:
```diff
@@ -62,15 +62,17 @@ FULL_VERSION    := $(VERSION)
+ifeq ($(ARCH), s390x)
+CONF_FILES	+= features.d/qeth.modules features.d/dasd_mod.modules
+_dummy		:= $(shell ${SED} -i -e 's:features=":features="qeth dasd_mod :g' mkinitfs.conf)
+endif
```
features.d/*.modules files could be added to CONF_FILES outside of this if-else block (the same place with others). But since we need to add to mkinitfs.conf anyway, I'd just add them all in this block.

ncopa reviewed (changes requested):
- don't use `sed -i` in makefile
- move `nameserver` to separate commit

tmh1999 committed:
initramfs-init: add disk and network detection on s390x

- Allow including dasd and qeth modules when building initramfs
- Add detection for dasd and qeth devices from cmdline in init

tmh1999 committed:
initramfs-init: add support for s390x installer

On s390x, the boot media consist of a kernel, an initramfs, and a parm
file. We don't necessarily need a modloop image and initial packages
but fetch from remote repo instead. This behavior is the same on other
distros.

The SSH server is required to allow root access with password during
installation. This config will be removed after running setup-disk.
User sets password in the parm file along with network interface
types, IP addresses, DNS server, disks.

ncopa commented on initramfs-init.in:
```diff
@@ -646,6 +652,19 @@ if [ -n "$ovlfiles" ]; then
+		rm $sysroot/.passwd
+	fi
+fi
+
```
As far I can see, there is nothing in this that is specific for s390x. Would it be an idea to rename the `s390x_installer` variable to `ssh_install` or similar?
